### ISSUE TITLE
OWASP rules for vacuum

### ISSUE THREAD

**Issue by Ricagraca** [daveshanley/vacuum#283]
Hello!

I was interested on running the owasp rules that exist for spectral https://github.com/stoplightio/spectral-owasp-ruleset/blob/main/src/ruleset.ts in vacuum. 

I am having some trouble translating some of the rules while debugging it from vacuum. Mostly regarding the Given "" statement and the function options. I have noticed that vacuum uses https://github.com/vmware-labs/yaml-jsonpath to parse Given. But I am still finding hard to translate (the first rule) 

e.g., `"$.paths..parameters[*][?(@property === "name" && (@ === "id" || @.match(/(_id|Id|-id)$/)))]^.schema"`. 

Mostly the parent  "^". I will continue working on it and see if I can make it to work (the tests provided by spectral will correspond to the ones on vacuum).

These are my questions: 
- Is there an easy way to translate these rules in order to guarantee the same semantics?
- Would people find it useful to have the owasp rules in vacuum? 

Thank you so much for your attention.

**Comment by daveshanley** [comment]
HI!

There has to be a way to translate the rules, and if there isn't, I can make it work one way or another.

The charged version of JSON Path that Spectral uses is a little frustrating at times, I know why they did what they did, but it really does complicate things downstream (like this issue). As you look at the regex, you're right that vacuum does depend on `yaml-path` which supports JSON Path syntax that is 'as close to the standard if there were a standard' and uses re2 syntax (https://github.com/google/re2/wiki/Syntax).

I would value OWASP support in vacuum, it makes a lot of sense, and it should be included out of the box. I've added this to the backlog.

**Comment by Ricagraca** [comment]
Super! Thanks for the quick response.

If you could provide me some hints on how could I approach this problem, I could help you extend it. 
At this moment I debugged vacuum with function options for the owasp:api1:2019-no-numeric-ids https://github.com/stoplightio/spectral-owasp-ruleset/blob/main/src/ruleset.ts:

```
functionOptions: {
  schema: {
    type: "object",
    not: {
      properties: {
        type: {
          const: "integer",
        },
      },
    },
    properties: {
      format: {
        const: "uuid",
      },
    },
  },
},
```

But the type const: "integer" is not being parsed correctly with 

```
func OWASP2() *model.Rule {

	// create a schema to match against.
	opts := make(map[string]interface{})
	yml := `type: object
not:
  properties:
    type:
      const: "integer"
properties:
  format:
    const: "uuid"`
(...)
```

I will continue on this and if possible I will open a PR so that  it can be looked at.

**Comment by daveshanley** [comment]
After reading through some of the rules, I am currently at the following conclusion. The reason why 'const: integer' is not picked up is that it's not a valid JSON Schema.

I could not find any other examples of using this design, and when I looked up the standard, I was reminded that 'type' can only be a string or an array of strings... it cannot be an object.

Here is the verbatim standard:

"The value of this keyword MUST be either a string or an array. If it is an array, elements of the array MUST be strings and MUST be unique.

String values MUST be one of the six primitive types ("null", "boolean", "object", "array", "number", or "string"), or "integer" which matches any number with a zero fractional part.

If the value of "type" is a string, then an instance validates successfully if its type matches the type represented by the value of the string. If the value of "type" is an array, then an instance validates successfully if its type matches any of the types indicated by the strings in the array."

https://json-schema.org/draft/2020-12/json-schema-validation.html#name-type

So, I don't know what to suggest. The rule looks bogus to me; it's not valid OpenAPI 3.1
The same applies to OpenAPI 3.0, 'type' can ONLY be a string, not an array, and not an object.

I don't know how this can ever be used in the wild.

Now I could be missing something completely, but just looking at the mechanics of the standards, there is no way this is a valid JSON Schema. 

So what to do next?

These rules can be implemented as first-class rules, built into vacuum. I think trying to implement them using the core functions is going to cause endless pain, particularly if the core rules are trying to do invalid things (like setting an object for a 'type' value).

The specific rule about determining if a parameter with a variation of 'id' uses a non-sequential value (like a UUID) is pretty simple to build using the existing design, as are all the others.

What would I do?

I would build new functions for each OWASP rule, create a new built-in ruleset for just applying the OWASP rules, and another option for 'extending' the recommended rules with the additional OWASP rules.

This way, all the logic for OWASP can be implemented directly into the core of the tool, and everyone 
gets that new powerful feature.

**Comment by Ricagraca** [comment]
Managed to create a PR for the first rule:
https://github.com/daveshanley/vacuum/pull/285

It seems that it is working fine.
In my free time I will continue to extend with the other rules and then, as you said, I can create a keyword for extends for the owasp stuff. Some feedback on it would be nice

**Comment by daveshanley** [comment]
OWASP rules have been available since `v0.2.0`. Read more here:

https://quobix.com/vacuum/rules/owasp/

and

https://quobix.com/vacuum/functions/owasp/

Thank you so very much to @Ricagraca for such great work on this feature.
